ls1intum/Artemis
# [Enhancement] Make group synchronization customizable for cypress tests

### Motivation and Context


In order to test programming exercises we need to dynamically create courses and programming exercises. Currently we have a limitation on servers that use bamboo and bitbucket that we have to wait at least 60 seconds between the creation of a course and a programming exercise. If we don't wait the creation of the progamming exercise fails. Currently the sleep is hardcoded so even if the Cypress tests are run against testservers, which use gitlab and jenkins, the tests still have a static sleep. To make testing on other testservers faster we should skip the group synchronization in this case.

### Description

This PR refactors the static sleeps and reduces code duplication by putting the logic into a custom cypress command. Furthermore a environmental parameter is added to tell Cypress if the target environment uses bamboo + bitbucket or gitlab + jenkins. If the parameter is false all group synchronization sleeps are skipped, so the tests should be 2-3 minutes faster than usual.

### Steps for Testing

Affected cypress tests:

- `ProgrammingExercise.spec.ts`
- `ProgrammingExerciseParticipation.spec.ts`
- `ExamManagement.spec.ts`

1. Configure cypress to run on ts2 with [our guide](https://confluence.ase.in.tum.de/display/ArTEMiS/Cypress+Tests)
2. Set the `isBamboo` parameter to `true` (this does not skip group synchronization)
3. Execute the affected cypress tests individually and note their total execution time (example is under screenshots)
4. Set the `isBamboo` parameter to `false` (this skips group synchronization)
5. Execute the affected cypress tests individually and compare the total execution time to the previous run
6. You should notice that the execution time has been reduced by about a minute for each test

### Review Progress





- Code Review

### Screenshots


#### With group synchronization
![image](https://user-images.githubusercontent.com/12861668/129630982-ef910cb6-36ef-4b04-a2d2-cda952c2d069.png)
#### Without group synchronization
![image](https://user-images.githubusercontent.com/12861668/129631200-37b29cc3-45cf-46eb-ad1f-a3585e9e6da3.png)

### Coverage
Bamboo plan (with group synchronization) <a href="https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-173">passes</a>
